src/typechecker.erl

```diff
@@ -2019,17 +2214,84 @@ type_check_rel_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
 	false ->
 	  throw({type_error, rel_error, Op, P, ResTy})
     end.
+
 type_check_list_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), list, []}, Env#env.tenv) of
-      {true, Cs} ->
-        {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-	{VarBinds2, Cs2} = type_check_expr_in(Env, ResTy, Arg2),
-	  {union_var_binds([VarBinds1, VarBinds2])
-	  ,constraints:combine([Cs, Cs1, Cs2])};
-      false ->
-	  throw({type_error, list_op_error, Op, P, ResTy})
+    Target =
+        %% '--' always produces a proper list, but '++' gives you an improper
+        %% list if the second argument is improper.
+        case Op of
+            '--' -> type(list, [type(term)]);
+            '++' -> type(maybe_improper_list, [type(term), type(term)])
+        end,
+    ResTy1 = glb(Target, ResTy, Env#env.tenv),
+    IsNone =
+        case ResTy1 of
+            {type, _, none, []} -> true;
+            _                   -> false
+        end,
+    case IsNone of
+        false ->
+            case list_op_arg_types(Op, ResTy1) of
+                {ArgTy1, ArgTy2} ->
+                    {VarBinds1, Cs1} = type_check_expr_in(Env, ArgTy1, Arg1),
+                    {VarBinds2, Cs2} = type_check_expr_in(Env, ArgTy2, Arg2),
+                    {union_var_binds([VarBinds1, VarBinds2]),
+                     constraints:combine([Cs1, Cs2])};
+                false ->
+                    throw({type_error, op_type_too_precise, Op, P, ResTy1})
+            end;
+        true ->
+            throw({type_error, list_op_error, Op, P, ResTy})
     end.
 
+list_op_arg_types('++', Ty) ->
+    case list_view(Ty) of
+        {empty, _, _}       -> {type(nil), type(nil)};
+        {Empty, Elem, _} ->
+            Arg1 = from_list_view({Empty, Elem, type(nil)}),
+            {Arg1, Ty}
```

Comment:
This is another instance of the `integer() + float() :: float()` issue. It would be nice to be able to check `list(A) ++ nonempty_list(A) :: nonempty_list(A)` and variants. Currently we push `nonempty_list` into both arguments.